crates/harp/src/environment_iter.rs

```diff
@@ -13,20 +11,17 @@ pub struct EnvironmentIter {
     names: std::vec::IntoIter<String>,
 }
 
-#[derive(Eq, PartialEq)]
 pub struct Binding {
     pub name: RSymbol,
     pub value: BindingValue,
 }
 
-// What is this used for? Do we still need it?
-// https://github.com/posit-dev/amalthea/commit/04a9fe20a48cb01c4fe2d6fe1a4dfdc0f2a186fc
-#[derive(Eq)]
-pub struct BindingNestedEnvironment {
-    pub has_nested_environment: bool,
+#[derive(Eq, PartialEq)]
+pub enum RObjectId {
+    Cons(SEXP, Box<RObjectId>),
```

Comment:
oh I did not realise the recursion would need to be indirected like that.

Now I wonder if a Vec would have been better so there's only one pointer dereference (to get the vector's array on the heap).

The other thing that now occurs to me is that the Id should be tagged by the original type to avoid potential ambiguity when two different wrappers of the same object have the same id structure (doesn't currently seem possible though).

I think it's fine as is though, just noting these potential issues. Maybe just add a comment that the id needs to be tagged if used in more complex situations.